text/3830-dedented-string-literals.md

```diff
@@ -0,0 +1,1569 @@
+
+   This is contrary to how we would normally write code, with indentation 1 level deeper than the surrounding.
+
+   ```diff
+   fn main() {
+       println!(
+           "\
+   +create table student(
+   +   id int primary key,
+   +   name text
+   +)");
+   }
+   ```
+
+   This makes it confusing to tell which scope the string belongs to. This is especially true when there are multile scopes involved:
+
+   ```rs
+   fn main() {
+       {
+           println!(
+               "\
+   create table student(
+      id int primary key,
+      name text
+   )");
+       }
+       println!(
+           "\
+   create table student(
+      id int primary key,
+      name text
+   )");
+       {
+           {
+               println!(
+                   "\
+   create table student(
+      id int primary key,
+      name text
+   )");
+            
+           } 
+       }
+   }
+   ```
+
+   All of the strings end up on the same level, despite them being in different scopes.
+
+3. The closing double-quote must be put at the beginning of the line, in order not to introduce trailing whitespace:
+
+   ```diff
+   fn main() {
+       println!(
+           "\
+   create table student(
+       id int primary key,
+       name text
+   +)");
+   }
+   ```
+
+As you can see, we have to choose one or the other. In either case we have to give something up.
+
+Sometimes, we are *forced* into the first option - sacrifice readability of the source.
+
+In some cases, producing excessive whitespace will change meaning of the output.
+
+Consider whitespace-sensitive languages such as Python or Haskell, or content which is meant to be read by people like generated Markdown - here we *can't* make a sacrifice on readabilty of the output - so our source code must become harder to understand.
+
+But, what if we could have the best of both worlds?
+
+### Dedented string literals
+
+In order to solve these problems, the RFC proposes dedented string literals of the form: `d"string"`.
+
+Common leading whitespace on each line after the closing quote in dedented string literals will be stripped at compile-time.
+
+This allows us to have a more readable version of the above:
+
+```rust
+fn main() {
+    println!(d"
+        create table student(
+            id int primary key,
+            name text
+        )
+        ");
+^^^^^^^^ // common leading whitespace (will be removed)
+}
+```
+
+All of the above problems are gracefully solved:
+
+1. Indentation level inside the string is the same as what is in the output.
+1. It does not require escaping the first newline for it to look readable.
+1. Nicely composes with raw string literal: `dr#"string"#`, in which the first newline *cannot* be escaped.
+1. Indentation level of the statement is larger than the `println!` call,
+   making it more obvious that the string is inside the call at a glance.
+1. The closing parentheses in the SQL statement aligs with `create table`
+   and is 1 level larger than `println!`.
```

Comment:
This seems like a downside to me. For example, compare:

```rs
my_function((
    a,
    b,
));
```

with:

```rs
my_function(d"
    a
    b
    ");
```

Seems a little inconsistent; the status quo is that closing brackets are idented to the level one above their contents, not the same level as their contents.